### PR TITLE
Make auto assignments of instance variables in templates configurable.

### PR DESCRIPTION
**Pull request by lorefnon**
As discussed in https://github.com/opal/opal-rails/issues/44#issuecomment-181123904 this provides a direct solution for "stack level too deep" error in case of json serialization failure of instance variables.
## Summary

When rendering opal templates, the auto assignment of instance variables can be disabled globally or on a per controller basis:

``` ruby

  # Globally disable:
  config.opal.auto_assign_instance_variables = false

  # Disable specific controllers (secondary) or specific actions (primary#without_assignments)
  config.opal.auto_assign_instance_variables = {
    except: ['primary#without_assignments', 'secondary']
  }

  # Enable specific controllers (secondary) or specific actions (primary#without_assignments)
  config.opal.auto_assign_instance_variables = {
    only: ['primary#without_assignments', 'secondary']
  }

```

### CODE REVIEW

File: lib/opal/rails/template_handler.rb [hunk 1]
```diff
@@ -23,7 +23,11 @@ def local_assigns
 
       def assigns
         <<-'RUBY'.strip
-          JSON.parse(@_assigns.to_json).map { |key, val| "@#{key} = #{val.inspect};" }.join
+          if Rails.application.config.opal.auto_assign_instance_variables?(params.slice(:controller, :action))
```
Comment (elia) [inline lib/opal/rails/template_handler.rb]: This `if` should be outside of the `<<-'RUBY'` heredoc string, even better is to use an early return:

```rb
unless Rails.application.config.opal.auto_assign_instance_variables?(params.slice(:controller, :action))

<<-'RUBY'.strip
…
```

**Comment by elia** [comment]
Thanks! 
Looks good generally, I have only a proposal on how to switch to a simpler way of handling the configuration and would like to hear your thoughts on that. 

What if instead of having a config entry we use a method api to check if a controller should auto assign variables?

```rb
def assigns
  return super if !respond_to?(:assign_opal_instance_variables?) || !assign_opal_instance_variables?

  <<-'RUBY'.strip
  …
  RUBY
end
```

That way the conditional can be anything ranging from `true`/`false` to a check in the database.

```rb
# ApplicationController
def assign_opal_instance_variables?
  action_name =~ /foo/
end
```

File: test_apps/rails5.rb [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 module RailsApp
   class Application < Rails::Application
-    config.root                                       = __dir__
+    config.root                                       = File.dirname(__FILE__)
```
Comment (elia) [inline test_apps/rails5.rb]: no need for this, Rails 5 is Ruby 2.3+ :)

File: test_apps/rails5.rb [hunk 3]
```diff
@@ -26,69 +26,25 @@ class Application < Rails::Application
...
     routes.append do
       get '/' => 'application#index'
-      get '/application/with_assignments' => 'application#with_assignments'
+      get '/primary/with_assignments' => 'primary#with_assignments'
+      get '/primary/without_assignments' => 'primary#without_assignments'
+      get '/secondary/without_assignments' => 'secondary#without_assignments'
     end
 
-    config.assets.paths << File.join(__dir__, 'assets/javascripts')
+    config.assets.paths << File.join(File.dirname(__FILE__), 'assets/javascripts')
```
Comment (elia) [inline test_apps/rails5.rb]: as above

File: lib/opal/rails/template_handler.rb (hunk 1 above)
Comment (lorefnon) [inline lib/opal/rails/template_handler.rb]: The problem with that in TemplateHandler context I do not have access to params. params becomes available only when the string is evaluated in view context. 

That is why I had added the the check inside the evaluated string. 

I am exploring the code around this first time, but if there is an alternative approach please let me know.

File: test_apps/rails5.rb (hunk 2 above)
Comment (lorefnon) [inline test_apps/rails5.rb]: In that case, would it be ok if I remove ruby 1.9 from travis config ? 

I was just worried that a red cross besides my PR would hamper the changes of it being accepted :)

**Comment by lorefnon** [comment]
Yes, the prospect of using a controller method is more flexible. I will update the PR with the relevant change.

**Comment by lorefnon** [comment]
@elia I have incorporated your suggestions.

**Comment by mibamur** [comment]
@elia maybe allow this and my PR to `develop` branch ;-)
@lorefnon hi!

and how about drop old rubies travis support?

**Comment by elia** [comment]
@lorefnon @mibamur sorry for the delay, reviewing this I got to the conclusion that I think it's better to have a simple global switch for pushing ivars/lvars to the template. 

Rails has already a way of letting the user limit what to pass down to the view by just not defining instance variables. I'm still interested in your consideration on this, in the meanwhile I'm closing this.

Added some form of credit in the changelog: https://github.com/opal/opal-rails/commit/bbdcea8107f48fa5c0434cb682dfea874ba8fecd – I hope it's appreciated 🙂

**Pull request closed without merging**